File: C++/find-x-value-of-array-i.cpp
// Time:  O(n * k)
// Space: O(k)

// dp
class Solution {
public:
    vector<long long> resultArray(vector<int>& nums, int k) {
        vector<long long> result(k);
        vector<long long> dp(k);
        for (const auto& x : nums) {
            vector<long long> new_dp(k);
            ++new_dp[x % k];
            for (int64_t i = 0; i < size(dp); ++i) {
                new_dp[i * x % k] += dp[i];
            }
            for (int i = 0; i < size(new_dp); ++i) {
                result[i] += new_dp[i];
            }
            dp = move(new_dp);
        }
        return result;
    }
};

File: Python/find-x-value-of-array-i.py
# Time:  O(n * k)
# Space: O(k)

# dp
class Solution(object):
    def resultArray(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        result = [0]*k
        dp = [0]*k
        for x in nums:
            new_dp = [0]*k
            new_dp[x%k] += 1
            for i, c in enumerate(dp):
                new_dp[i*x%k] += c
            for i, c in enumerate(new_dp):
                result[i] += c 
            dp = new_dp
        return result
